### PR TITLE
Add aggregate function combinator -Resample

### PR DESCRIPTION
**Pull request by hczhcz**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
- New Feature

Short description (up to few sentences):

`-Resample(begin,end,step)(..., key)` can be appended to an aggregate function.The aggregate function will have an extra `key` argument, and it will run separately for rows which have keys in each slot `[begin, begin + step), [begin + step, begin + 2 * step), ...`.

Detailed description (optional):

Example:
```
|uid   |age   |
|'A'   |16    |
|'B'   |30    |
|'C'   |35    |
|'D'   |48    |
|'E'   |62    |
```
```sql
SELECT groupArrayResample(0, 80, 10)(uid, age)
    = [[], ['A'], [], ['B', 'C'], ['D'], [], ['E'], []];
```

### CODE REVIEW

File: dbms/src/AggregateFunctions/AggregateFunctionResample.h [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+            AggregateFunctionResample<Key>
+        > {arguments, params},
+        nested_function {nested_function},
+        last_col {arguments.size() - 1},
+        begin {begin},
+        end {end},
+        step {step},
+        total {
+            static_cast<size_t>(
+                (end - begin + step - (step >= 0 ? 1 : -1)) / step
```
Comment (hczhcz) [inline dbms/src/AggregateFunctions/AggregateFunctionResample.h]: TODO: check zero-division manually

**Comment by hczhcz** [comment]
TODO: Fix param handling

**Comment by alesapin** [comment]
`select arrayReduce('sumResample(1, 1, 0)', [10, 11, 12, 13, 14, 15, 16, 17, 18, 19], [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]);` leads to Floating point exception and server crash https://pastebin.com/3UzexYKW.

File: dbms/src/AggregateFunctions/AggregateFunctionResample.cpp [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+                nested_function,
+                params.front().get<Int64>(),
+                params[1].get<Int64>(),
+                params.back().get<Int64>(),
+                arguments,
+                params
+            );
+
+        // TODO
+        return nullptr;
```
Comment (alesapin) [inline dbms/src/AggregateFunctions/AggregateFunctionResample.cpp]: It's ok to throw an exception here.

File: dbms/src/AggregateFunctions/AggregateFunctionResample.h (hunk 1 above)
Comment (alesapin) [reply dbms/src/AggregateFunctions/AggregateFunctionResample.h]: Yes, it's still have to be checked.

**Comment by alesapin** [comment]
What is the main idea of the negative step and intervals? I don't understand how they work:
```
SELECT arrayReduce('sumResample(0, -1, -1)', [10], [0])

┌─arrayReduce('sumResample(0, -1, -1)', [10], [0])─┐
│ []                                               │
└──────────────────────────────────────────────────┘
SELECT arrayReduce('sumResample(0, -1, -1)', [10], [1])

┌─arrayReduce('sumResample(0, -1, -1)', [10], [1])─┐
│ []                                               │
└──────────────────────────────────────────────────┘
SELECT arrayReduce('sumResample(9, -1, -1)', [10], [0])                                                                                                                                                                                                                                                
                                                                                                                                                                                                                                                                                                       
┌─arrayReduce('sumResample(9, -1, -1)', [10], [0])─┐                                                                                                                                                                                                                                                   
│ [0,0,0,0,0,0,0,0,0,10]                           │                                                                                                                                                                                                                                                   
└──────────────────────────────────────────────────┘   
SELECT arrayReduce('sumResample(0, -1, -1)', [10], [-0])                                                                                                                                                                                                                                                
                                                                                                                                                                                                                                                                                                       
┌─arrayReduce('sumResample(0, -1, -1)', [10], [-0])─┐                                                                                                                                                                                                                                                   
│ [10]                                             │                                                                                                                                                                                                                                                   
└──────────────────────────────────────────────────┘   
```

**Review by alesapin — CHANGES_REQUESTED** [review]
.

**Comment by alesapin** [comment]
Also error messages are unclear:
```
SELECT arrayReduce('sumResample(11, 10, 1)', [10], [0])

Received exception from server (version 19.9.1):
Code: 36. DB::Exception: Received from localhost:9000, ::1. DB::Exception: The range given in function sumResample contains too many elements. 
```

**Comment by YoannBuch** [comment]
Hello @hczhcz, very nice addition :)

How would one use `-Resample` in combination of `quantileTimingWeighted`? (knowing that `quantileTimingWeighted` already returns an array) Thanks!

**Pull request merged**